OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorSelectionLogic.cs

```diff
@@ -0,0 +1,160 @@
+		public MapEditorSelectionLogic(Widget widget, World world, WorldRenderer worldRenderer)
+		{
+			this.worldRenderer = worldRenderer;
+
+			editorActorLayer = world.WorldActor.Trait<EditorActorLayer>();
+			resourceLayer = world.WorldActor.Trait<IResourceLayer>();
+
+			editor = widget.Get<EditorViewportControllerWidget>("MAP_EDITOR");
+
+			var selectTabContainer = widget.Get("SELECT_WIDGETS");
+
+			actorEditPanel = selectTabContainer.Get<ContainerWidget>("ACTOR_EDIT_PANEL");
+			areaEditPanel = selectTabContainer.Get<ContainerWidget>("AREA_EDIT_PANEL");
+
+			actorEditPanel.IsVisible = () => editor.CurrentBrush == editor.DefaultBrush && editor.DefaultBrush.Selection.Actor != null;
+			areaEditPanel.IsVisible = () => !actorEditPanel.IsVisible();
+
+			copyTerrainCheckbox = areaEditPanel.Get<CheckboxWidget>("COPY_FILTER_TERRAIN_CHECKBOX");
+			copyResourcesCheckbox = areaEditPanel.Get<CheckboxWidget>("COPY_FILTER_RESOURCES_CHECKBOX");
+			copyActorsCheckbox = areaEditPanel.Get<CheckboxWidget>("COPY_FILTER_ACTORS_CHECKBOX");
+
+			var copyButton = widget.Get<ButtonWidget>("COPY_BUTTON");
+			copyButton.OnClick = () => clipboard = CopySelectionContents();
+
+			copyButton.IsDisabled = () => editor.DefaultBrush.Selection.Area == null;
+
+			var pasteButton = widget.Get<ButtonWidget>("PASTE_BUTTON");
+			pasteButton.OnClick = () =>
+			{
+				if (clipboard == null)
+					return;
+
+				editor.SetBrush(new EditorCopyPasteBrush(
+					editor,
+					worldRenderer,
+					clipboard,
+					resourceLayer,
+					() => copyFilters));
+			};
+
+			pasteButton.IsDisabled = () => clipboard == null;
+			pasteButton.IsHighlighted = () => editor.CurrentBrush is EditorCopyPasteBrush;
+
+			var closeAreaSelectionButton = areaEditPanel.Get<ButtonWidget>("SELECTION_CANCEL_BUTTON");
+			closeAreaSelectionButton.OnClick = () => editor.DefaultBrush.ClearSelection();
+
+			CreateCategoriesPanel();
+		}
+
+		EditorClipboard CopySelectionContents()
+		{
+			var selection = editor.DefaultBrush.Selection.Area;
+			var source = new CellCoordsRegion(selection.TopLeft, selection.BottomRight);
+
+			var mapTiles = worldRenderer.World.Map.Tiles;
+			var mapHeight = worldRenderer.World.Map.Height;
+			var mapResources = worldRenderer.World.Map.Resources;
+
+			var previews = new Dictionary<string, EditorActorPreview>();
+			var tiles = new Dictionary<CPos, ClipboardTile>();
+
+			foreach (var cell in source)
+			{
+				if (!mapTiles.Contains(cell))
+					continue;
+
+				var offset = new CVec(selection.TopLeft.X, selection.TopLeft.Y);
+
+				var resourceLayerContents = resourceLayer.GetResource(cell);
+
+				tiles.Add(cell, new ClipboardTile
+				{
+					TerrainTile = mapTiles[cell],
+					ResourceTile = mapResources[cell],
+					ResourceLayerContents = resourceLayerContents,
+					Height = mapHeight[cell]
+				});
+
+				if (copyFilters.HasFlag(MapCopyFilters.Actors))
+				{
+					var regionActors = selection.SelectMany(editorActorLayer.PreviewsAt).Distinct().ToList();
```

Comment:
```suggestion
					var regionActors = selection.SelectMany(editorActorLayer.PreviewsAt).ToHashSet();
```

This does the same except faster